### PR TITLE
userlib: Enable support for rv64

### PR DESCRIPTION
**Pull request by furquan-rivos**
This change copies over riscv32.rs in sys/userlib/src/arch to riscv64.rs and updates it accordingly for rv64 architecture.

### CODE REVIEW

**Review by adurbin-rivos — APPROVED** [review]

File: sys/userlib/src/arch/riscv64.rs [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+        lw a0, 6*8(a0)    # packed_target_operation
+
+        # Load the constant syscall number.
+        li a7, {sysnum}
+
+        # To the kernel!
+        ecall
+
+        slli a1, a1, 0x20
+        or a0, a0, a1
```
Comment (FawazTirmizi) [inline sys/userlib/src/arch/riscv64.rs:96]: Why was this added here and at lines 212 and 243?

**Review by FawazTirmizi — APPROVED** [review]
LGTM, but can you get rid of the merge commit before merging? I think it's best to keep them out of our histories where possible.

**Comment by furquan-rivos** [comment]
> LGTM, but can you get rid of the merge commit before merging? I think it's best to keep them out of our histories where possible.

github takes care of that when doing a rebase and merge into base branch. We wouldn't see that in the base branch history.

File: sys/userlib/src/arch/riscv64.rs (hunk 1 above)
Comment (furquan-rivos) [reply sys/userlib/src/arch/riscv64.rs:96]: The kernel places `resp` in ret0 and `len` in ret1 when returning back from syscall. Within userlib arch-agnostic code, `sys_send()` expects RcLen as return value from `sys_send_stub()` which is provided by arch-dependent code. RcLen is a struct with a single `u64` type. `sys_send()` takes this RcLen and calls `from()` to convert it into `(u32, usize)`.

So, on 32-bit arch, the `from()` expects RcLen to be held in ret0 and ret1. However, on 64-bit arch, the `from()` expects RcLen to be held in a single ret0. That is why RV64 is applying the transformation to put together the two 32-bit ret values into a single u64.

This is what `from()` looks like on rv64:
```
000000009002311e <from>:
    9002311e:   02055593                srli    a1,a0,0x20
    90023122:   8082                    ret
```

and this is what it looks like on rv32:
```
200150b2 <from>:
200150b2:       8082                    ret
```

Comment (furquan-rivos) [reply sys/userlib/src/arch/riscv64.rs:96]: I think at some point we should switch to passing RcLen as an argument into `sys_send_stub()` and fill that in from arch-dependent code. Then we can just address individual members directly instead of having to do this weird shifting in arch-dependent code. Much difficult to follow. Also, it restricts us to use len which is 32-bit only on 64-bit arch.

Comment (furquan-rivos) [reply sys/userlib/src/arch/riscv64.rs:96]: Raised issue [SW-560](http://go/ji/SW-560).
